src/providers/auth.ts

```diff
@@ -0,0 +1,59 @@
+import {Provider, Inject} from 'angular2/core';
+import {ReplaySubject} from 'rxjs/subject/ReplaySubject';
+import {FirebaseRef} from '../tokens';
+
+import * as Firebase from 'firebase';
+
+const kBufferSize = 1; 
+
+export enum AuthProviders {
+  Github
+};
+
+export class Auth<T extends ProviderMetadata> extends ReplaySubject<AuthState<T>> {
+  constructor (@Inject(FirebaseRef) private _fbRef: Firebase) {
+    super (kBufferSize);
+
+    this._fbRef.onAuth((authData) => this._emitAuthData(authData));
+  }
+
+  private _emitAuthData(authData: any): void {
+    if (authData == null) {
+      this.next(null);
+    } else {
+      let {uid, provider, github} = authData;
+      let providerMetadata: T;
+      switch (provider) {
+        case "github":
+          providerMetadata = github;
+          provider = AuthProviders.Github;
+          break;
+        default:
+          throw new Error(`Unsupported firebase auth provider ${provider}`);
+      }
+      this.next({
+        uid,
+        provider,
+        providerMetadata
+      });
+    }
+  }
+}
+
+export interface AuthState <T extends ProviderMetadata > {
+  uid: string;
+  provider: AuthProviders;
+  providerMetadata?: T;
+}
+
+export interface GithubAuthMetadata extends ProviderMetadata {
+  id: string;
+  username: string;
+  accessToken: string;
+  displayName: string;
```

Comment:
Most metadata fields can change based on scope, so we should only type the basics and leave the rest to the `cachedUserProfile` object.